core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala

```diff
@@ -876,24 +898,36 @@ class TimeSeriesShard(val dataset: Dataset,
 
   private[memstore] val addPartitionsDisabled = AtomicBoolean(false)
 
+  //scalastyle:off
   private[filodb] def getOrAddPartition(recordBase: Any, recordOff: Long, group: Int, ingestOffset: Long) = {
-    val part = partSet.getOrAddWithIngestBR(recordBase, recordOff, {
-      val partKeyOffset = recordComp.buildPartKeyFromIngest(recordBase, recordOff, partKeyBuilder)
-      val newPart = createNewPartition(partKeyArray, partKeyOffset, group)
-      if (newPart != OutOfMemPartition) {
-        val partId = newPart.partID
-        // NOTE: Don't use binRecordReader here.  recordOffset might not be set correctly
-        val startTime = dataset.ingestionSchema.getLong(recordBase, recordOff, timestampColId)
-        partKeyIndex.addPartKey(newPart.partKeyBytes, partId, startTime)()
-        timeBucketBitmaps.get(currentIndexTimeBucket).set(partId)
-        activelyIngesting.set(partId)
-        logger.trace(s"Created new partition ${newPart.stringPartition} on dataset=${dataset.ref} " +
-          s"shard $shardNum at offset $ingestOffset")
-      }
-      newPart
-    })
-    part
+    partSet.getWithIngestBR(recordBase, recordOff) match {
+      case null =>
+        val stamp = partSetLock.writeLock()
+        try {
+          val part = partSet.getOrAddWithIngestBR(recordBase, recordOff, {
```

Comment:
I understand why the double check is there.  What I'm suggesting is to replace the second `getOrAddWithIngestBR` with simply `add`, because the check in `getOrAdd` is superfluous -- we already did the check with the first get call.  After all `getOrAdd` is simply the get followed by an add.

Reply:
That makes sense, although what I'm doing is double-checked locking. Although no other thread is likely to call this method, the second check prevents any odd issues.

Reply:
I'm going fix the code such that an optimistic read lock is used first.